tests/scenario/test_extension_providers.py

```diff
@@ -0,0 +1,203 @@
+    data = {"a": 1, "b": ["1", "2", "3"], "c": ["1", "2"]}
+    prefix = "provider[1]"
+    expected = {
+        "provider[1].a": 1,
+        "provider[1].b[0]": "1",
+        "provider[1].b[1]": "2",
+        "provider[1].b[2]": "3",
+        "provider[1].c[0]": "1",
+        "provider[1].c[1]": "2",
+    }
+    result = flatten_config(data, prefix)
+    assert result == expected
+
+
+def test_flatten_config_with_tuple_values():
+    """Verify that tuple values are flattened into individual elements with indices when a prefix is provided."""
+    data = {"x": (10, 20), "y": ("a", "b", "c")}
+    prefix = "provider"
+    expected = {
+        "provider.x[0]": 10,
+        "provider.x[1]": 20,
+        "provider.y[0]": "a",
+        "provider.y[1]": "b",
+        "provider.y[2]": "c",
+    }
+    result = flatten_config(data, prefix)
+    assert result == expected
+
+
+def test_flatten_config_with_empty_prefix_flat_data():
+    """Verify that flattening flat data structures with an empty prefix produces keys without a leading dot."""
+    data = {"a": 1, "b": ["1", "2", "3"], "c": ["1", "2"]}
+    prefix = ""
+    expected = {
+        "a": 1,
+        "b[0]": "1",
+        "b[1]": "2",
+        "b[2]": "3",
+        "c[0]": "1",
+        "c[1]": "2",
+    }
+    result = flatten_config(data, prefix)
+    assert result == expected
+
+
+def test_flatten_config_with_empty_prefix_nested_data():
+    """Verify that flattening nested data structures with an empty prefix produces keys without a leading dot."""
+    data = {"a": {"b": 1}, "c": ["1", "2", "3"], "d": {"e": ["1", "2", "3"]}}
+    prefix = ""
+    expected = {
+        "a.b": 1,
+        "c[0]": "1",
+        "c[1]": "2",
+        "c[2]": "3",
+        "d.e[0]": "1",
+        "d.e[1]": "2",
+        "d.e[2]": "3",
+    }
+    result = flatten_config(data, prefix)
+    assert result == expected
+
+
+def test_flatten_config_primitive():
+    """Verify that a primitive value (non-dict, non-list, non-tuple) is directly assigned to the prefix key."""
+    value = 42
+    prefix = "number"
+    expected = {"number": 42}
+    result = flatten_config(value, prefix)
+    assert result == expected
+
+
+def test_flatten_config_primitive_no_prefix():
+    """Verify that a primitive value with an empty prefix is returned with an empty key."""
+    value = "primitive"
+    prefix = ""
+    expected = {"": "primitive"}
+    result = flatten_config(value, prefix)
+    assert result == expected
+
+
+#############################################
+# Tests for _build_extension_providers_config
+#############################################
+
+
+def test_build_extension_providers_config_empty(istio_core_context):
+    """Test that an empty providers list returns an empty configuration."""
+    state = State(relations=[])
+    with istio_core_context(istio_core_context.on.config_changed(), state) as mgr:
```

Comment:
nit: since these tests dont use anything from the event handling, I'd use update-status (since we do less work on those events and its less likely to interfere).  But this works too